### DISCUSSION TITLE
Duplicates with cape-capf-super

### DISCUSSION THREAD

**Discussion by grolongo** [minad/cape#104] [Q&A]
Hi minad,

Since 029d1a6, I'm getting duplicates for each candidate with `yasnippet-capf`.

It is only happening when I'm using `cape-capf-super` though, as in:

```elisp
(defun my/cape-emacs-lisp ()
  (setq-local completion-at-point-functions
              (list (cape-capf-super
                     #'yasnippet-capf))))

(add-hook 'emacs-lisp-mode-hook 'my/cape-emacs-lisp)
```

![Screenshot (531)](https://github.com/minad/cape/assets/34292770/06a2f26c-3161-461c-b0bc-2c07d51b604e)

From your earlier commit messages, I read that duplicates should be handled by the completion UI. Does it mean I have to fix that through `corfu` or `yasnippet-capf`?

Thanks for the help.

**Comment by minad** [comment]
Thanks for the report. This mean that `yasnippet-capf` produces duplicates. The easiest solution is to make sure that backends to not produce unwanted duplicates. I can reproduce the problem with this test code:

~~~ elisp
(setq completion-at-point-functions (list (cape-capf-super #'test-capf)))
(setq completion-at-point-functions (list #'test-capf))

(defun test-capf ()
  (when-let ((bounds (bounds-of-thing-at-point 'symbol)))
    (list (car bounds) (cdr bounds) '("first" "first" "second" "third"))))
~~~

Corfu displays duplicate candidates if they differ in their text properties. This leads to the problem here since `cape-capf-super` has to manipulate the candidates to distinguish them, leading to the unwanted duplication here, see:

https://github.com/minad/cape/blob/f021305d20a40021058fdd31f9acddc86e463eaf/cape.el#L907-L910

We should improve the handling of duplicates in `cape-capf-super`. The difficulty here is to preserve duplicates for the different backends (they should make it through). Try this:

~~~ elisp
(setq completion-at-point-functions (list (cape-capf-super #'test-capf #'test-capf)))
~~~

There you see that duplicates (two "second" for example) from the two backends make it through. This is desired since different backends may create different effects when you select the different candidates. Unfortunately what we lost on the way for `cape-capf-super` is that duplicates of a single backend are still deduplicated. We also should do that efficiently. But again, the simpler solution might be if backends combined by `cape-capf-super` do not produce duplicates in the first place.

**Reply by grolongo** [reply]
I agree, if two different backends produce identical candidates both should be able to display it, and users could get hinted from which backend both candidates belong to:
```
foo (tempel)
foo (dabbrev)
```

I tested your examples above using your latest commit and it works as intended, no duplicates showing.

From what you've said though it seems that it's the backend itself that is producing duplicates and therefore is the culprit. Why is `#'yasnippet-capf` producing dups but `#'cape-abbrev` and others aren't? Isn't it something that should be fixed in `yasnippet-capf` directly instead of `cape`?

**Comment by minad** [comment]
I think I managed to improve the deduplication in https://github.com/minad/cape/commit/249f6702e99343aac045697178e319784aacd2f0. I am not entirely sure if all scenarios are fixed. The test cases here work properly now:

~~~ elisp

(defun test-capf1 ()
  (when-let ((bounds (bounds-of-thing-at-point 'symbol)))
    (list (car bounds) (cdr bounds) '("first" "first" "second" "third" "fourth"))))

(defun test-capf2 ()
  (when-let ((bounds (bounds-of-thing-at-point 'symbol)))
    (list (car bounds) (cdr bounds) (mapcar
                                     (lambda (x)
                                       (propertize x 'face 'error))
                                     '("first" "second" "third" "fourth")))))

;; Test case: Single capf with duplicates -> Duplicates removed
(setq completion-at-point-functions (list #'test-capf1))

;; Test case: Super + Single capf with duplicates -> Duplicates removed
(setq completion-at-point-functions (list (cape-capf-super #'test-capf1)))

;; Test case: Super + two times the same capf -> Duplicates removed
;; Note that sorting by the UI is enabled, since otherwise the candidates would not get deduplicated.
(setq completion-at-point-functions
      (list
       (cape-capf-properties (cape-capf-super #'test-capf1 #'test-capf1) :sort t)))

;; Test case: Super + two different capfs -> Duplicates preserved
;; Note that sorting by the UI is enabled, since otherwise the candidates would not get deduplicated.
(setq completion-at-point-functions
      (list
       (cape-capf-properties (cape-capf-super #'test-capf1 #'test-capf2) :sort t)))
~~~

**Reply by grolongo** [reply]
I tested the fourth examples with the latest commit and I get the same results as predicted in the comments. Good.

Now I tested again with `yasnippet-capf`, and I get a weird behavior:

Let's say I have a snippet **foobar** and I set `corfu-auto-prefix` to `1`.
When I type **f** I get **foobar** duplicated in the candidate results.
If I type **o** after the first **f** (fo), the duplicate disappears.
If I hit backspace to delete the **o**, the duplicate reappears.

Now when I set `corfu-auto-prefix` to `2`, everything works fine.

Maybe this snippet will give you less of a headache:

```elisp
(setq-local completion-at-point-functions
            (list (cape-capf-super
                   #'yasnippet-capf)))

(setq corfu-auto-prefix 1)

;; this shows foobar twice (duplicates) in corfu
f|
;; this removes the duplicate and foobar appears only once (good)
fo|
;; now I hit backspace and the duplicate reappears
f|

(setq corfu-auto-prefix 2)

;; no unintended duplicates whatsoever
fo|
```

As I mentioned in the comment from your first answer, this is only happening with
`yasnippet-capf`, not with other backends.

**Reply by minad** [reply]
Thanks for this. If you manage to produce a minimal example with test capfs I can look into this again. I don't have the capacity to debug code which I don't maintain. I suspect that the `yasnippet-capf` prefilters candidates in some way, based on the prefix. This is unnecessary and the backend should not do this, since otherwise the prefiltering interferes with the completion style filtering, and also the deduplication it seems. The simple backends `tempel-complete` and `cape-keyword` operate in a similar way and can be used as reference.

**Reply by grolongo** [reply]
>If you manage to produce a minimal example with test capfs I can look into this again.  

I tried several different backends combined with and without `cape-capf-super` and couldn't produce a single unwanted duplicate, so that's a good thing for cape :)

>I don't have the capacity to debug code which I don't maintain.  

No problem at all, at first I only wanted to know if this was something that should be fixed in corfu/cape because of 029d1a6 or yasnippet-capf. Now thanks to all your testing and good explanations I understand it better and I know where to fix it. I'll try to have someone reproduce dups from the yasnippet-capf repository with `corfu-auto-prefix 1`. Thank you a lot for your help.

**Comment by minad** [comment]
grolongo ***@***.***> writes:

> I agree, if two different backends produce identical candidates both should be
> able to display it, and users could get hinted from which backend both
> candidates belong to:
>
> foo (tempel)
> foo (dabbrev)

Yes, this would be neat. Right now, cape-capf-super doesn't override the
annotations, but maybe it should.

> From what you've said though it seems that it's the backend itself that is
> producing duplicates and therefore is the culprit. Why is #'yasnippet-capf
> producing dups but #'cape-abbrev and others aren't? Isn't it something that
> should be fixed in yasnippet-capf directly instead of cape?

The specification of the Capf mechanism allows backends to create
duplicates, which should be filtered out, except if they look
differently or produce different effects. This means the behavior is
most complicated, some duplicates should be kept, some should be thrown
away.

Allowing duplicates makes sense, for example the duplicates of
cape-capf-super produce different effects. Therefore we have to support
duplicates and as a consequence the spec would be better if the frontend
would not have to do any duplicate filtering at all. I would prefer such
a simpler specification, where backends were not allowed to produce
unwanted duplicates.

I just recently implemented this feature. Earlier, Corfu always filtered
out all `equal` duplicates. Now Corfu uses `equal-including-properties`
to distinguish candidates which still make a difference.

In some cases unwanted duplicates without different effects are produced
by the backend. Sometimes this makes a backend easier to implement.
Nevertheless I consider it a bug, careless implementation or at least an
inefficiency if a backend produces such unwanted duplicates. For the
Cape backends I make sure that no unwanted duplicates are generated in
the first place.

**Comment by jeff-phil** [comment]
> grolongo ***@***.***> writes:
 >> I agree, if two different backends produce identical candidates both should be
 >>  able to display it, and users could get hinted from which backend both
 >> candidates belong to:
 >> 
 >> foo (tempel)
 >> foo (dabbrev)
 >>
> Yes, this would be neat. Right now, cape-capf-super doesn't override the
> annotations, but maybe it should.

Yes, I always have to know where things are coming from with completions.   I add annotations like this to any capf's that I combine via `cape-capf-super`.  I normally just do it in the super definition as shown, but for tempel one I did it separate since I wanted to merge the label with existing annotation that expands.  Here's my example for elisp completions:

``` elisp
  ;; tempel has a useful annotation-func that shows template, so this adds to it.
  (defalias 'my/capf-tempel
    (let* ((old-properties (cdr (cape-wrap-properties #'tempel-complete)))
           (old-ann-fun (plist-get old-properties :annotation-function)))
      (cape-capf-properties
       #'tempel-complete
       :annotation-function
       #'(lambda (cand)
           (concat (funcall `,old-ann-fun cand) " "
                   (propertize " Tempel"
                               'face font-lock-comment-face))))))
  (defun my/cape-tempel (&optional interactive)
    (interactive (list t))
    (when interactive
      (when (bound-and-true-p copilot-mode) (copilot-clear-overlay))
      (cape-interactive #'my/capf-tempel)))
  (add-to-list 'completion-at-point-functions #'my/cape-tempel)
  (keymap-global-set "C-c p /" #'my/cape-tempel)

  (defalias 'my/capf-elisp+et-al
    (cape-capf-super
     ;; This allows a completion-function to return if codeium is off, vs nil error
     (cape-capf-silent
      (when (bound-and-true-p my/codeium-is-enabled)
        (cape-capf-properties #'codeium-completion-at-point
                              :annotation-function
                              #'(lambda (_) (propertize "  Codium"
                                                        'face font-lock-comment-face))
                              :company-kind (lambda (_) 'magic))))
     #'cape-elisp-symbol
     #'my/capf-tempel
     ;;cape-line too chatty, this quiets a bit
     (cape-capf-silent
      (cape-capf-properties #'cape-line
                            :annotation-function
                            #'(lambda (_) (propertize " Line"
                                                      'face font-lock-comment-face))))
     (cape-capf-properties #'cape-dabbrev
                           :annotation-function
                           #'(lambda (_) (propertize " Dabbrev"
                                                     'face font-lock-comment-face)))))
  (defun my/cape-elisp+et-al (&optional interactive)
    (interactive (list t))
    (when interactive
      (when (bound-and-true-p copilot-mode) (copilot-clear-overlay))
      (cape-interactive #'my/capf-elisp+et-al)))
  (keymap-global-set "C-c p x" #'my/cape-elisp+et-al)
```

Here's what it ultimately looks like:

<img width="567" alt="image" src="https://github.com/minad/cape/assets/6047093/addaf08a-adcb-42ce-9c89-e42a4785da0c">
<br/><br/>

For the super completion, it uses normal `completion-at-point`, `TAB` -> `corfu-complete` key binding.  But, I also have separate key bindings for discreet options.  Those are the other extra bindings you see.

[Other notes, I was experimenting with codeium and copilot.  Codeium leverages capf, so it gets added into the super if enabled.  And copilot uses overlay and not capf, so if that is enabled then clear the overlay before starting completion-at-point. I left that in above, but feel free to zap it.]
